resource/file/fetch/fetch.go

```diff
@@ -0,0 +1,217 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fetch
+
+import (
+	"crypto/md5"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/hex"
+	"fmt"
+	"hash"
+	"io"
+	"net/url"
+	"os"
+
+	"github.com/asteris-llc/converge/resource"
+	"github.com/hashicorp/go-getter"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+)
+
+// Hash type for Fetch
+type Hash string
+
+const (
+	// HashMD5 indicates hash type md5
+	HashMD5 Hash = "md5"
+
+	// HashSHA1 indicates hash type sha1
+	HashSHA1 Hash = "sha1"
+
+	// HashSHA256 indicates hash type sha256
+	HashSHA256 Hash = "sha256"
+
+	// HashSHA512 indicates hash type sha512
+	HashSHA512 Hash = "sha512"
+)
+
+// Fetch gets a file and makes it available on disk
+type Fetch struct {
+	Source      string
+	Destination string
+	HashType    string
+	Hash        string
+	Force       bool
+
+	hasApplied bool
+}
+
+// Check if the the file is on disk, and the hashes are availabe
+func (f *Fetch) Check(context.Context, resource.Renderer) (resource.TaskStatus, error) {
+	var (
+		hsh    hash.Hash
+		err    error
+		status = resource.NewStatus()
+	)
+
+	if !f.hasApplied {
+		if f.Hash != "" {
+			hsh, err = f.getHash()
+			if err != nil {
+				status.RaiseLevel(resource.StatusCantChange)
+				return status, errors.Wrap(err, "will not attempt file fetch")
+			}
+		}
+
+		status, err = f.DiffFile(status, hsh)
+		if err != nil {
+			return status, err
+		}
+	}
+
+	return status, nil
+}
+
+// Apply fetches the file
+func (f *Fetch) Apply(context.Context) (resource.TaskStatus, error) {
+	var (
+		hsh      hash.Hash
+		err      error
+		status   = resource.NewStatus()
+		checksum = ""
+	)
+
+	if f.Hash != "" {
+		hsh, err = f.getHash()
+		if err != nil {
+			status.RaiseLevel(resource.StatusCantChange)
+			return status, errors.Wrap(err, "will not attempt file fetch")
+		}
+	}
+
+	stat, err := f.DiffFile(status, hsh)
+	if err != nil {
+		return stat, err
+	} else if resource.AnyChanges(stat.Differences) {
+		u, err := url.Parse(f.Source)
+		if err != nil {
+			status.RaiseLevel(resource.StatusFatal)
+			return status, errors.Wrap(err, "could not parse source")
+		}
+
+		values := u.Query()
+		values.Set("archive", "false")
+		if f.Hash != "" {
+			checksum = fmt.Sprintf("checksum=%s:%s", f.HashType, f.Hash)
+		}
+		source := u.String() + "?" + values.Encode() + checksum
+
+		pwd, err := os.Getwd()
+		if err != nil {
+			status.RaiseLevel(resource.StatusFatal)
+			return status, errors.Wrap(err, "failed to get working directory")
+		}
+
+		client := &getter.Client{
+			Src:  source,
+			Dst:  f.Destination,
+			Pwd:  pwd,
+			Mode: getter.ClientModeFile,
+		}
+		if err := client.Get(); err != nil {
+			status.RaiseLevel(resource.StatusFatal)
+			return status, errors.Wrap(err, "failed to fetch")
+		}
+		status.AddMessage("fetched successfully")
+		f.hasApplied = true
+	}
+
```

Comment:
I wonder if it might be worth it here to actually set `f.Hash` even if the user hadn't originally specified one- the idea being that someone might want to do a lookup on the hash of a fetched file even if they didn't actually specify an explicit one.  What do you think?